Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,467 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOHTTP2
+import NIOTLS
+
+/// A connection pool that manages and creates new connections to hosts respecting the specified preferences
+class ConnectionPool {
+    /// The default look group used to schedule async work
+    let loopGroup: EventLoopGroup
+
+    /// The configuration used to bootstrap new HTTP connections
+    let configuration: HTTPClient.Configuration
+
+    /// The main data structure used by the `ConnectionPool` to retreive and create connections associated
+    /// to a a given `Key` .
+    /// - Warning: This property shouldn't be directly accessed, use the `ConnectionPool` subscript instead
+    var _connectionProviders: [Key: ConnectionProvider] = [:]
```

Comment:
are there any reasons for this (and other properties) not be private?

Reply:
Thanks! AFAICT even the unit tests probably won't ever need this so this can probably be private instead of internal.

Reply:
also, suuuper nitpicky, maybe we should simplify the name to just `providers`? and `connectionProvidersLock` to just `lock`?